File: B_3666_求数列所有后缀最大值的位置.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define int unsigned long long

signed main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    
    int n, ans = 0;
    cin >> n;
    vector<int> a(n + 1);
    stack<int> s;
    for (int i = 1; i <= n; ++i) {
        cin >> a[i];
        while (s.size() and a[i] >= a[s.top()]) {
            ans ^= s.top();
            s.pop();
        }
        ans ^= i;
        s.emplace(i);
        cout << ans << endl;
    }
    return 0;
}

File: B_3666_求数列所有后缀最大值的位置.py
n = int(input())
ans = 0
a = [0] + list(map(int, input().split()))
s = []
for i in range(1, n + 1):
    while s and a[i] > a[s[-1]]:
        ans ^= s[-1]
        s.pop()
    ans ^= i
    s.append(i)
    print(ans)
